src/Microsoft.AspNetCore.Razor.Evolution/DefaultDirectiveIRPass.cs

```diff
@@ -48,13 +66,40 @@ public override void VisitDirective(DirectiveIRNode node)
                 }
                 else if (string.Equals(node.Name, CSharpCodeParser.InheritsDirectiveDescriptor.Name, StringComparison.Ordinal))
                 {
+                    node.Parent.Children.Remove(node);
+
                     var token = node.Tokens.FirstOrDefault();
 
                     if (token != null)
                     {
                         _classNode.BaseType = token.Content;
                     }
                 }
+                else if (string.Equals(node.Name, CSharpCodeParser.SectionDirectiveDescriptor.Name, StringComparison.Ordinal))
+                {
+                    var sectionIndex = node.Parent.Children.IndexOf(node);
+                    node.Parent.Children.Remove(node);
+
+                    var defineSectionEndStatement = new CSharpStatementIRNode()
+                    {
+                        Content = "});",
+                    };
+                    node.Parent.Children.Insert(sectionIndex, defineSectionEndStatement);
+
+                    foreach (var child in node.Children.Except(node.Tokens).Reverse())
+                    {
+                        node.Parent.Children.Insert(sectionIndex, child);
+                    }
+
+                    var lambdaContent = _designTime ? "__razor_section_writer" : string.Empty;
+                    var sectionName = node.Tokens.FirstOrDefault()?.Content;
+                    var defineSectionStartStatement = new CSharpStatementIRNode()
+                    {
+                        Content = /* ORIGINAL: DefineSectionMethodName */ $"DefineSection(\"{sectionName}\", async ({lambdaContent}) => {{",
+                    };
+
+                    node.Parent.Children.Insert(sectionIndex, defineSectionStartStatement);
```

Comment:
I could make a custom node that acted as a container and had 3 children. Thoughts?